internal/match/dict.go

```diff
@@ -0,0 +1,405 @@
+// Copyright 2020 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package match defines matching algorithms and support code for the license checker.
+package match
+
+import (
+	"strings"
+	"unicode"
+	"unicode/utf8"
+)
+
+// A Dict maps words to integer indexes in a word list, of type WordID.
+// The zero Dict is an empty dictionary ready for use.
+//
+// Lookup and Words are read-only operations,
+// safe for any number of concurrent calls from multiple goroutines.
+// Insert is a write operation; it must not run concurrently with
+// any other call, whether to Insert, Lookup, or Words.
+type Dict struct {
+	dict map[string]WordID // dict maps word to index in words
+	list []string          // list of known words
+}
+
+// A WordID is the index of a word in a dictionary.
+type WordID int32
+
+// BadWord represents a word not present in the dictionary.
+const BadWord WordID = -1
+
+// Insert adds the word w to the word list, returning its index.
+// If w is already in the word list, it is not added again; Insert returns the existing index.
+func (d *Dict) Insert(w string) WordID {
+	id, ok := d.dict[w]
+	if ok {
+		return id
+	}
+	if d.dict == nil {
+		d.dict = make(map[string]WordID)
+	}
+	id = WordID(len(d.list))
+	if int(id) != len(d.list) {
+		panic("dictionary too large")
+	}
+	d.list = append(d.list, w)
+	d.dict[w] = id
+	return id
+}
+
+// Lookup looks for the word w in the word list and returns its index.
+// If w is not in the word list, Lookup returns BadWord.
+func (d *Dict) Lookup(w string) WordID {
+	id, ok := d.dict[w]
+	if !ok {
+		return BadWord
+	}
+	return id
+}
+
+// Words returns the current word list.
+// The list is not a copy; the caller can read but must not modify the list.
+func (d *Dict) Words() []string {
```

Comment:
the contract could be enforced by  having the returned object be a struct containing an unexported slice, if you care.